package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -0,0 +1,690 @@
+
+        cdef bint memory_error = False
+
+        with nogil:
+            for i in range(size_search):
+                if memory_error:
+                    break
+                current_beadid = search_ids_view[i]
+                cellindex = self.grid.cellids[current_beadid]
+                self.grid.cellid2cellxyz(cellindex, cellxyz)
+                for xi in range(DIM):
+                    if memory_error:
+                        break
+                    for yi in range(DIM):
+                        if memory_error:
+                            break
+                        for zi in range(DIM):
+                            if memory_error:
+                                break
+                            # Calculate and/or reinitialize shifted coordinates
+                            shifted_coord[XX] = self.coords[current_beadid, XX] + (xi - 1) * self.grid.cellsize[XX]
+                            shifted_coord[YY] = self.coords[current_beadid, YY] + (yi - 1) * self.grid.cellsize[YY]
+                            shifted_coord[ZZ] = self.coords[current_beadid, ZZ] + (zi - 1) * self.grid.cellsize[ZZ]
+                            probe[XX] = self.coords[current_beadid, XX] + (xi - 1) * self.cutoff
+                            probe[YY] = self.coords[current_beadid, YY] + (yi - 1) * self.cutoff
+                            probe[ZZ] = self.coords[current_beadid, ZZ] + (zi - 1) * self.cutoff
+                            # Make sure the shifted coordinates is inside the brick-shaped box
+                            for m in range(DIM - 1, -1, -1):
+                                while shifted_coord[m] < 0:
+                                    for d in range(m+1):
+                                        shifted_coord[d] += self.box.c_pbcbox.box[m][d]
+                                while shifted_coord[m] >= self.box.c_pbcbox.box[m][m]:
+                                    for d in range(m+1):
+                                        shifted_coord[d] -= self.box.c_pbcbox.box[m][d]
+                                while probe[m] < 0:
+                                    for d in range(m+1):
+                                        probe[d] += self.box.c_pbcbox.box[m][d]
+                                while probe[m] >= self.box.c_pbcbox.box[m][m]:
+                                    for d in range(m+1):
+                                        probe[d] -= self.box.c_pbcbox.box[m][d]
+
+                            # Get the cell index corresponding to the coord
+                            cellindex_adjacent = self.grid.coord2cellid(shifted_coord)
+                            cellindex_probe = self.grid.coord2cellid(probe)
+
+                            for j in range(self.grid.nbeads[cellindex_adjacent]):
+                                bid = self.grid.beadids[cellindex_adjacent * self.grid.nbeads_per_cell + j]
+                                if checked[bid] != 0:

```

Comment:
I think it could be better to have the checked flag on the cell rather than every coordinate in a cell. Since every cell is an entity as opposed to ever atom in cell-lists. AFAIK, every time we need to do the evaluation, we need to check atleast one cell (all coordinates in the cell). 

Here, cell_index probe might become useful to reduce the the computations since the cutoff radius is always smaller than or equal to the cellsize. To prove my point, consider a box with size 10, cellsize 2 in all direction. For a cutoff distance of 0.4 and at the centre of any cell, all the shifts with the probe will still land the coordinates in the same cell, and if the checkcell is present, we just need to compute the distances once and not check every coordinate individually.